src/Workspaces/Core/Portable/Workspace/Workspace_EventsLegacy.cs

```diff
@@ -0,0 +1,140 @@
+    /// </summary>
+    public event EventHandler<TextDocumentEventArgs> TextDocumentOpened
+    {
+        add => AddEventHandler(value, WorkspaceEventType.TextDocumentOpened);
+        remove => RemoveEventHandler(value, WorkspaceEventType.TextDocumentOpened);
+    }
+
+    /// <summary>
+    /// An event that is fired when a <see cref="Document"/> is closed in the editor.
+    /// </summary>
+    public event EventHandler<DocumentEventArgs> DocumentClosed
+    {
+        add => AddEventHandler(value, WorkspaceEventType.DocumentClosed);
+        remove => RemoveEventHandler(value, WorkspaceEventType.DocumentClosed);
+    }
+
+    /// <summary>
+    /// An event that is fired when any <see cref="TextDocument"/> is closed in the editor.
+    /// </summary>
+    public event EventHandler<TextDocumentEventArgs> TextDocumentClosed
+    {
+        add => AddEventHandler(value, WorkspaceEventType.TextDocumentClosed);
+        remove => RemoveEventHandler(value, WorkspaceEventType.TextDocumentClosed);
+    }
+
+    /// <summary>
+    /// An event that is fired when the active context document associated with a buffer 
+    /// changes.
+    /// </summary>
+    public event EventHandler<DocumentActiveContextChangedEventArgs> DocumentActiveContextChanged
+    {
+        add => AddEventHandler(value, WorkspaceEventType.DocumentActiveContextChanged);
+        remove => RemoveEventHandler(value, WorkspaceEventType.DocumentActiveContextChanged);
+    }
+
+    private void AddEventHandler<TEventArgs>(EventHandler<TEventArgs> eventHandler, WorkspaceEventType eventType)
+        where TEventArgs : EventArgs
+    {
+        // Require main thread on the callback as this is used from publicly exposed events
+        // and those callbacks may have main thread dependencies.
+        IDisposable? disposer = null;
+
+        lock (_gate)
+        {
+            if (_disposableEventHandlers.TryGetValue((eventHandler, eventType), out var adviseCountAndDisposer))
+            {
+                (var adviseCount, disposer) = adviseCountAndDisposer;
+
+                // If we already have a handler for this event type, update the map with the new advise count
+                _disposableEventHandlers[(eventHandler, eventType)] = (adviseCount + 1, disposer);
+            }
+        }
+
+        if (disposer == null)
+        {
+            disposer = RegisterHandler(eventType, (Action<EventArgs>)Handler, WorkspaceEventOptions.RequiresMainThreadOptions);
+
+            lock (_gate)
+            {
+                _disposableEventHandlers[(eventHandler, eventType)] = (AdviseCount: 1, disposer);
+            }
+        }
+
+        void Handler(EventArgs arg)
+            => eventHandler(sender: this, (TEventArgs)arg);
+    }
+
+    private void RemoveEventHandler<TEventArgs>(EventHandler<TEventArgs> eventHandler, WorkspaceEventType eventType)
+        where TEventArgs : EventArgs
+    {
+        var existingAdviseCount = 0;
```

Comment:
Consider just assigning to disposer only if it needs to be disposed; it removes the local.